server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -216,6 +221,23 @@ public void addNamedQuery(String name, Query query) {
         return Map.copyOf(namedQueries);
     }
 
+    public ParsedDocument parseDocument(SourceToParse source) throws MapperParsingException {
+        return mapperService.documentMapper() == null ? null : mapperService.documentMapper().parse(source);
+    }
+
+    public FieldNameAnalyzer getFieldNameIndexAnalyzer() {
+        DocumentMapper documentMapper = mapperService.documentMapper();
+        return documentMapper == null ? null : documentMapper.mappers().indexAnalyzer();
+    }
+
+    public boolean hasNested() {
+        return mapperService.hasNested();
+    }
+
+    public boolean hasMappings() {
+        return mapperService.documentMapper() != null;
+    }
```

Comment:
This is only used for the RandomScoreFunctionBuilder, correct?  Do we need to add it, or can we just remove that 'no mappings' special case?

Reply:
yes that is surprisingly the only usage. but I don't feel comfortable removing that no mappings special case. I think it's legit, although annoying to support, and we test it.

Reply:
@romseygeek are you ok with this or maybe do you have other ideas?

Reply:
Let's leave it in for now, if we decide to go with a no-op mapper by default in MapperService we can remove it then.

Reply:
true, I will try that idea very soon.